### DISCUSSION TITLE
DB2 connection issue with hammerdb

### DISCUSSION THREAD

**Discussion by Prathyusha6** [TPC-Council/HammerDB#154] [Q&A] [answered]
Hi @sm-shaw ,

**Facing this error :**
loading history file ... 0 events added
Main console display active (Tcl8.6.9 / Tk8.6.9)
The xml is well-formed, applying configuration
Error in Virtual User 1: [IBM][CLI Driver] SQL10007N Message "0" could not be retrieved. Reason code: "3".

**Details:**
DB2 version: 11.5.4.0
driver: go driver (tried with ODBC driver as well, facing the same issue)

**env vars:**
```
export DB2HOME=/home/user/go/src/github.com/ibmdb/go_ibm_db/installer/clidriver
export CGO_CFLAGS=-I$DB2HOME/include
export CGO_LDFLAGS=-L$DB2HOME/lib
export LD_LIBRARY_PATH=/home/user/go/src/github.com/ibmdb/go_ibm_db/installer/clidriver/lib
export PATH=$PATH:/home/user/go/src/github.com/ibmdb/go_ibm_db/installer/clidriver/bin
```

created **db2cli.ini** file

```
[tstcli1x]
uid=####
pwd=####
autocommit=0
TableType="'TABLE','VIEW','SYSTEM TABLE'"

[tstcli2x]
; Assuming dbalias2 is a database in DB2 for MVS.
SchemaList="'OWNER1','OWNER2',CURRENT SQLID"

[MyVeryLongDBALIASName]
dbalias=dbalias3
SysSchema=MYSCHEMA
```
ran sample test.go file successfully as follows (to check the connection from driver to DB2):

```
package main
import (
    _ "github.com/ibmdb/go_ibm_db"
    "database/sql"
    "fmt"
)
func main(){
    con:="HOSTNAME=host;DATABASE=name;PORT=number;UID=username;PWD=password"
    db, err:=sql.Open("go_ibm_db", con)
    if err != nil{
        fmt.Println(err)
    }
    err = db.Ping()
    if err != nil {
    fmt.Println(err)
    } else {
     fmt.Println("CONNECTION SUCCESS")
    }
    db.Close()
}
```
Hammerdb-3.3 is passing librarycheck for DB2. DB2 instance creation and sample database creation has been successful. But somehow hammerdb is not able to communicate with db2. May I know where I am going wrong here, please?

Thanks,
Prathyusha.

**Comment by sm-shaw** [comment]
Hi,
It looks like you are trying to use a driver for the GO language? HammerDB is written in the TCL language (https://www.hammerdb.com/blog/uncategorized/hammerdb-concepts-and-architecture/) and uses the db2tcl interface that in turn uses the Db2 CLI driver.  The error looks like it could be in your library path. Make sure that the required Db2 client library libdb2.so.1 in the lib64 directory is first in your library path. Then run the hammerdcli librarycheck command. You can also use ldd on the HammerDB db2tcl library to check the paths and make sure all dependencies are found. If this works the library loads successfully. Then make sure you configure your db2dsdriver.cfg and db2cli.ini files and you should be good to connect to your Db2 database. 
Steve

**Comment by Prathyusha6** [comment]
Hi @sm-shaw ,

Thanks for the reply. Yeah I removed go driver. And installed cli driver.

Here is **db2dsdriver.cfg**

![image](https://user-images.githubusercontent.com/12022780/95726682-dbf7a600-0c96-11eb-84ee-3f27b332ab25.png)


This libdb2.so.1 file is already there in installed directory should I replace, this file with client driver's libdb2.so.1 
opt/dbpath/lib32/libdb2.so.1
opt/dbpath/lib64/libdb2.so.1

db2tcl librarycheck is successfully passing in hammerdbcli command. 

The error still prevails, Am using RHEL 7.5
Kernel:  3.10.0-957.el7.x86_64 2018 x86_64 x86_64 x86_64 GNU/Linux

Also tried running db2val :

![image](https://user-images.githubusercontent.com/12022780/95731740-8377d700-0c9d-11eb-97af-d4b0ad917948.png)

db2val and db2sampl working fine in DB2 instance. But not as root in the server.
checked if this is cache problem, but mem/cache looks fine.

![image](https://user-images.githubusercontent.com/12022780/95732006-d487cb00-0c9d-11eb-919e-c9274132956a.png)


Any possibilities for OS incompatibilities as well?
And also am running both Hammerdb and DB2 on the same linux machine, is this fine?



Thanks,
Prathyusha

**Comment by sm-shaw** [comment]
Yes running HammerDB on the same machine as the database is fine. Here is an example from a test system. 
The LD_LIBRARY_PATH puts the lib64 directory first. 
running the ldd command shows that the correct library is detected, librarycheck works as does connecting to the database. 
```
echo $LD_LIBRARY_PATH
/home/steve/sqllib/lib64:/home/steve/sqllib/lib64/gskit:/home/steve/sqllib/lib32

ldd libdb2tcl.so.0.0.1 
	linux-vdso.so.1 (0x00007ffd159ce000)
	libdb2.so.1 => /home/steve/sqllib/lib64/libdb2.so.1 (0x00007fd1c66c4000
```

**Comment by sm-shaw** [comment]
This link may also help:
https://www.ibm.com/support/pages/node/154915

**Comment by Prathyusha6** [comment]
Hi @sm-shaw ,

yeah, db2start is working fine using DB2INSTANCE env var.

```
echo $DB2INSTANCE
db2inst5 
```

ldd libdb2tcl.so.0.0.1 is giving the folowing result:

![image](https://user-images.githubusercontent.com/12022780/95744839-a613eb00-0cb1-11eb-8a88-ea4e09ec1bd7.png)


```
 echo $LD_LIBRARY_PATH
/home/db2inst5/sqllib/lib64:/home/db2inst5/sqllib/lib64/gskit:/home/db2inst5/sqllib/lib32
```

libtcl.8.6.so is not found, this is causing issue ?

Thanks,
Prathyusha

**Comment by Prathyusha6** [comment]
Hi @sm-shaw ,

Please guide me if I'm doing this wrong:

am giving 
DB2 USER- instance username
DB2 pwd - instance password

Should I give something else other than these ?

![image](https://user-images.githubusercontent.com/12022780/95774399-fa34c480-0cdd-11eb-8894-9f37324404ce.png)

Thanks,
Prathyusha

**Comment by sm-shaw** [comment]
In the listing the libtcl library is missing. I don't see that in my test. Are you running hammerdb as the Db2 user or as root? It should be the db2 user.The IBM link tends to suggest a permission issue.

**Comment by Prathyusha6** [comment]
Hi @sm-shaw ,

Yeah, I have fixed libtcl library issue. 
Also, ldd libdb2tcl.so.0.0.1 command running only inside the directory of Hammerdb/lib/db2tcl, is this fine? 

I was trying as a root to run Hammerdb, so i have installed hammerdb again inside the instance so that i can run it as a db2 user. But I am facing an issue of DISPLAY env variable. Tried exporting DISPLAY=:0.0 but still the GUI command (./hammerdb) is giving the error. Is there any other way to run this. 
To debug this DISPLAY issue tried all the things like doing xhost+ and all. Still not able to debug. Can you please let me know if there is any other way or if I'm missing anything here.

Also tried through CLI as db2 user
![image](https://user-images.githubusercontent.com/12022780/95899443-36375a80-0dae-11eb-9fd4-f3e058df8f4c.png)

It's not working here as well. 
Thanks,
Prathyusha

**Comment by sm-shaw** [comment, marked as answer]
We need to separate the Linux DISPLAY configuration from the HammerDB one. DISPLAY configuration is not something for HammerDB to help with. 
For the error message it continues to look very much like a configuration/path issue - the error indicates that it cannnot find the correct error message in the catalog. I recommend rechecking the installation and configuration.

**Comment by Prathyusha6** [comment]
Thanks a lot for the support steve. 
It worked as a root only for me. Needed to change the directory of Hammerdb to /home, and also needed to add host ip and hostname in /etc/hosts.  :) 

Thanks,
Prathyusha.

**Comment by sm-shaw** [comment]
Many thanks for the update and confirming that the issue is resolved
